tensorflow_probability/python/experimental/vi/cascading_flows.py

```diff
@@ -0,0 +1,607 @@
+            variables=dist._model_unflatten(
+                # pylint: disable=protected-access
+                _extract_variables_from_coroutine_model(
+                    posterior_generator, seed=seed)))
+
+    # Temporary workaround for bijector caching issues with autobatched JDs.
+    surrogate_posterior = joint_distribution_auto_batched.JointDistributionCoroutineAutoBatched(
+        posterior_generator,
+        use_vectorized_map=dist.use_vectorized_map,
+        name=_get_name(dist))
+
+    # Ensure that the surrogate posterior structure matches that of the prior.
+    # todo: check me, do we need this? in case needs to be modified
+    #  if we use auxiliary variables, then the structure won't match the one of the
+    #  prior
+    '''try:
+      tf.nest.assert_same_structure(dist.dtype, surrogate_posterior.dtype)
+    except TypeError:
+      tokenize = lambda jd: jd._model_unflatten(
+        # pylint: disable=protected-access, g-long-lambda
+        range(len(jd._model_flatten(jd.dtype)))
+        # pylint: disable=protected-access
+      )
+      surrogate_posterior = restructure.Restructure(
+        output_structure=tokenize(dist),
+        input_structure=tokenize(surrogate_posterior))(
+        surrogate_posterior, name=_get_name(dist))'''
+    return surrogate_posterior, variables
+
+
+# todo: sample_shape is not used.. can remove?
+def _cf_convex_update_for_base_distribution(dist,
+                                            initial_prior_weight,
+                                            num_auxiliary_variables,
+                                            num_layers,
+                                            global_auxiliary_variables,
+                                            variables,
+                                            sample_shape=None,
+                                            seed=None):
+    """Creates a trainable surrogate for a (non-meta, non-joint) distribution."""
+
+    if variables is None:
+        actual_event_shape = dist.event_shape_tensor()
+        int_event_shape = int(actual_event_shape) if \
+            actual_event_shape.shape.as_list()[0] > 0 else 1
+        bijectors = [reshape.Reshape([-1],
+                                     event_shape_in=actual_event_shape +
+                                                    num_auxiliary_variables)]
+
+        bijectors.extend(
+            _build_highway_flow_block(
+                num_layers,
+                width=tf.reduce_prod(
+                    actual_event_shape + num_auxiliary_variables),
+                residual_fraction_initial_value=initial_prior_weight,
+                gate_first_n=int_event_shape, seed=seed))
+
+        bijectors.append(
+            reshape.Reshape(actual_event_shape + num_auxiliary_variables))
+
+        variables = chain.Chain(bijectors=list(reversed(bijectors)))
+
+    if num_auxiliary_variables > 0:
+        batch_shape = global_auxiliary_variables.shape[0] if len(
```

Comment:
the workaround 
`  if tf.nest.is_nested(prior_batch_shape):
      prior_batch_shape = functools.reduce(tf.broadcast_static_shape,
                                           dist._model_flatten(prior_batch_shape))`
seems to fail for models such as brownian and lorenz, where there is no batch dimension.

The code works instead for `test_shapes_model()` (even though using `constraining_bijector = dist.experimental_default_event_space_bijector()` gives problems with such model).

Do you have suggestions on how to make the code work in both cases, with and without batch shapes? (maybe a simple `try except` could be enough).
Also, regarding `constraining_bijector`, could the problem be with the `WishartTriL` distribution? In case, we could probably think about a test that has batch shape but doesn't use `WishartTriL`